### PR TITLE
Adding AppInsights ConnectionString support.

### PR DESCRIPTION
**Pull request by v-iyamauchi**
## Description
Adding AppInsights ConnectionString support as requested by a customer. The PR includes,
- Application Insights package upgrade
- Azure Monitor Otel Distro package upgrade (beta => release)
- Add a support for enabling AppInsights telemetry with a connection string.
- Add log enrichers for Otel telemetry logging.

## Related issues
Addresses [issue #104859].
[User Story 104859] (https://microsofthealth.visualstudio.com/Health/_workitems/edit/104859): Application Insights instrumentation keys are being deprecated (GitHub Issue #3376)

## Testing
Tested by running E2E tests and adding UTs.

## FHIR Team Checklist
- **Update the title** of the PR to be succinct and less than 65 characters
- **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- Tag the PR with the type of update: **Bug**, **Build**, **Dependencies**, **Enhancement**, **New-Feature** or **Documentation**
- Tag the PR with **Open source**, **Azure API for FHIR** (CosmosDB or common code) or **Azure Healthcare APIs** (SQL or common code) to specify where this change is intended to be released.
- Tag the PR with **Schema Version backward compatible** or **Schema Version backward incompatible** or **Schema Version unchanged** if this adds or updates Sql script which is/is not backward compatible with the code.
- [ ] CI is green before merge [Build Status](https://microsofthealthoss.visualstudio.com/FhirServer/_build/latest?definitionId=27&branchName=main) 
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs [hunk 1]
```diff
@@ -22,91 +27,158 @@
...
+
+            IServiceCollection services = new ServiceCollection();
+            Startup startup = new Startup(configuration);
+            var addTelemetryProviderMethod = startup.GetType().GetMethod(AddTelemetryProviderMethodName, BindingFlags.NonPublic | BindingFlags.Instance);
+            addTelemetryProviderMethod.Invoke(startup, new object[] { services });
+
+            Assert.Empty(services.Where(descriptor =>
+                descriptor.ImplementationType != null ? descriptor.ImplementationType.FullName.Contains("ApplicationInsights", StringComparison.OrdinalIgnoreCase) : false));
+            Assert.Empty(services.Where(descriptor =>
+                descriptor.ImplementationType != null ? descriptor.ImplementationType.FullName.Contains("OpenTelemetry", StringComparison.OrdinalIgnoreCase) : false));
```
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs [hunk 2]
```diff
@@ -22,91 +27,158 @@
...
+                configureServiceOptions?.Configure(serviceOptions);
+#pragma warning disable CS0618 // Type or member is obsolete
+                Assert.True(
+                    !string.IsNullOrWhiteSpace(instrumentationKey)
+                        ? serviceOptions.InstrumentationKey == instrumentationKey && serviceOptions.ConnectionString == null
+                        : serviceOptions.ConnectionString == connectionString && serviceOptions.InstrumentationKey == null);
+#pragma warning restore CS0618 // Type or member is obsolete
+
+                var configureTelemetryConfiguration = provider.GetRequiredService<IConfigureOptions<TelemetryConfiguration>>();
+                var telemetryConfiguration = new TelemetryConfiguration();
```
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web/AzureMonitorOpenTelemetryLogEnricher.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                AddOperationName(newAttributes);
+                data.Attributes = newAttributes.ToList();
+            }
+
+            base.OnEnd(data!);
+        }
+
+        public void AddOperationName(Dictionary<string, object?> attributes)
+        {
+            var operationName = string.Empty;
```
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web/AzureMonitorOpenTelemetryLogEnricher.cs]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web.UnitTests/AzureMonitorOpenTelemetryLogEnricherTests.cs [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+        }
+
+        [Fact]
+        public void GivenRequest_WhenOperationNameIsAbsent_ThenOperationNameShouldBeAddedWithMethodAndPath()
+        {
+            _httpContext.Request.RouteValues.ReturnsNull();
+            _httpContext.Request.Method = "GET";
+            _httpContext.Request.Path = $"/{Guid.NewGuid()}/health/check";
+
+            var operationName = $"{_httpContext.Request.Method} {_httpContext.Request.Path}";
```
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web.UnitTests/AzureMonitorOpenTelemetryLogEnricherTests.cs:51]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web/AzureMonitorOpenTelemetryLogEnricher.cs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+                        var parameterArray = request.RouteValues.Keys?.Where(
+                            k => k.Contains(KnownHttpRequestProperties.RouteValueParameterSuffix, StringComparison.OrdinalIgnoreCase))
+                            .OrderBy(k => k, StringComparer.OrdinalIgnoreCase)
+                            .ToArray();
+                        if (parameterArray != null && parameterArray.Any())
+                        {
+                            name += $" [{string.Join("/", parameterArray)}]";
+                        }
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web/AzureMonitorOpenTelemetryLogEnricher.cs]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web/Startup.cs [hunk 6]
```diff
@@ -243,36 +277,91 @@
...
+                                    var parameterArray = request.RouteValues.Keys?.Where(
+                                        k => k.Contains(KnownHttpRequestProperties.RouteValueParameterSuffix, StringComparison.OrdinalIgnoreCase))
+                                        .OrderBy(k => k, StringComparer.OrdinalIgnoreCase)
+                                        .ToArray();
+                                    if (parameterArray != null && parameterArray.Any())
+                                    {
+                                        name += $" [{string.Join("/", parameterArray)}]";
+                                    }
+                                }
+                            }
```
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web/Startup.cs]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs [hunk 7]
```diff
@@ -22,91 +27,158 @@
...
-            addOpenTelemetryMethod.Invoke(startup, new object[] { services.Object });
-            Assert.NotEmpty(serviceDescriptors.Where(descriptor =>
+            IConfiguration configuration = BuildConfiguration(TelemetryProviderNoneConfigurationValue, null, null);
+            IServiceCollection services = new ServiceCollection();
+            Startup startup = new Startup(configuration);
+            var addTelemetryProviderMethod = startup.GetType().GetMethod(AddTelemetryProviderMethodName, BindingFlags.NonPublic | BindingFlags.Instance);
+            addTelemetryProviderMethod.Invoke(startup, new object[] { services });
+
+            Assert.Empty(services.Where(descriptor =>
+                descriptor.ImplementationType != null ? descriptor.ImplementationType.FullName.Contains("ApplicationInsights", StringComparison.OrdinalIgnoreCase) : false));
```
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs [hunk 8]
```diff
@@ -22,91 +27,158 @@
...
+        public void GivenAppSettings_WhenConnectionInfoIsMissing_ThenTelemetryProviderShouldBeDisabled()
+        {
+            IConfiguration configuration = BuildConfiguration(TelemetryProviderApplicationInsightsConfigurationValue, null, null);
+            IServiceCollection services = new ServiceCollection();
+            Startup startup = new Startup(configuration);
+            var addTelemetryProviderMethod = startup.GetType().GetMethod(AddTelemetryProviderMethodName, BindingFlags.NonPublic | BindingFlags.Instance);
+            addTelemetryProviderMethod.Invoke(startup, new object[] { services });
+
+            Assert.Empty(services.Where(descriptor =>
+                descriptor.ImplementationType != null ? descriptor.ImplementationType.FullName.Contains("ApplicationInsights", StringComparison.OrdinalIgnoreCase) : false));
```
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs [hunk 9]
```diff
@@ -22,91 +27,158 @@
...
+            IServiceCollection services = new ServiceCollection();
+            Startup startup = new Startup(configuration);
+            var addTelemetryProviderMethod = startup.GetType().GetMethod(AddTelemetryProviderMethodName, BindingFlags.NonPublic | BindingFlags.Instance);
+            addTelemetryProviderMethod.Invoke(startup, new object[] { services });
+
+            using var provider = services.BuildServiceProvider();
+            if (!string.IsNullOrWhiteSpace(instrumentationKey) || !string.IsNullOrWhiteSpace(connectionString))
+            {
+                Assert.NotEmpty(services.Where(descriptor =>
+                    descriptor.ImplementationType != null ? descriptor.ImplementationType.FullName.Contains("ApplicationInsights", StringComparison.OrdinalIgnoreCase) : false));
```
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs [hunk 10]
```diff
@@ -22,91 +27,158 @@
...
+                Assert.True(
+                    !string.IsNullOrWhiteSpace(instrumentationKey)
+                        ? telemetryConfiguration.InstrumentationKey == instrumentationKey && string.IsNullOrEmpty(telemetryConfiguration.ConnectionString)
+                        : telemetryConfiguration.ConnectionString == connectionString);
+#pragma warning restore CS0618 // Type or member is obsolete
+            }
+            else
+            {
+                Assert.Empty(services.Where(descriptor =>
+                    descriptor.ImplementationType != null ? descriptor.ImplementationType.FullName.Contains("ApplicationInsights", StringComparison.OrdinalIgnoreCase) : false));
```
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs [hunk 11]
```diff
@@ -22,91 +27,158 @@
...
+            IServiceCollection services = new ServiceCollection();
+            Startup startup = new Startup(configuration);
+            var addTelemetryProviderMethod = startup.GetType().GetMethod(AddTelemetryProviderMethodName, BindingFlags.NonPublic | BindingFlags.Instance);
+            addTelemetryProviderMethod.Invoke(startup, new object[] { services });
+
+            using var provider = services.BuildServiceProvider();
+            if (!string.IsNullOrWhiteSpace(connectionString))
+            {
+                Assert.NotEmpty(services.Where(descriptor =>
+                    descriptor.ImplementationType != null ? descriptor.ImplementationType.FullName.Contains("OpenTelemetry", StringComparison.OrdinalIgnoreCase) : false));
```
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs [hunk 12]
```diff
@@ -22,91 +27,158 @@
...
+            }
+            else
+            {
+                Assert.Empty(services.Where(descriptor =>
+                    descriptor.ImplementationType != null ? descriptor.ImplementationType.FullName.Contains("OpenTelemetry", StringComparison.OrdinalIgnoreCase) : false));
+            }
+
+            Assert.Empty(services.Where(descriptor =>
+                descriptor.ImplementationType != null ? descriptor.ImplementationType.FullName.Contains("ApplicationInsights", StringComparison.OrdinalIgnoreCase) : false));
```
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs [hunk 13]
```diff
@@ -22,91 +27,158 @@
...
+                        ? telemetryConfiguration.InstrumentationKey == instrumentationKey && string.IsNullOrEmpty(telemetryConfiguration.ConnectionString)
+                        : telemetryConfiguration.ConnectionString == connectionString);
+#pragma warning restore CS0618 // Type or member is obsolete
+            }
+            else
+            {
+                Assert.Empty(services.Where(descriptor =>
+                    descriptor.ImplementationType != null ? descriptor.ImplementationType.FullName.Contains("ApplicationInsights", StringComparison.OrdinalIgnoreCase) : false));
+                Assert.Empty(services.Where(descriptor =>
+                    descriptor.ImplementationType != null ? descriptor.ImplementationType.FullName.Contains("OpenTelemetry", StringComparison.OrdinalIgnoreCase) : false));
```
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs [hunk 14]
```diff
@@ -22,91 +27,158 @@
...
+            using var provider = services.BuildServiceProvider();
+            if (!string.IsNullOrWhiteSpace(connectionString))
+            {
+                Assert.NotEmpty(services.Where(descriptor =>
+                    descriptor.ImplementationType != null ? descriptor.ImplementationType.FullName.Contains("OpenTelemetry", StringComparison.OrdinalIgnoreCase) : false));
+            }
+            else
+            {
+                Assert.Empty(services.Where(descriptor =>
+                    descriptor.ImplementationType != null ? descriptor.ImplementationType.FullName.Contains("OpenTelemetry", StringComparison.OrdinalIgnoreCase) : false));
```
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs [hunk 15]
```diff
@@ -22,91 +27,158 @@
...
+            var addTelemetryProviderMethod = startup.GetType().GetMethod(AddTelemetryProviderMethodName, BindingFlags.NonPublic | BindingFlags.Instance);
+            addTelemetryProviderMethod.Invoke(startup, new object[] { services });
+
+            using var provider = services.BuildServiceProvider();
+            if (!string.IsNullOrWhiteSpace(instrumentationKey) || !string.IsNullOrWhiteSpace(connectionString))
+            {
+                Assert.NotEmpty(services.Where(descriptor =>
+                    descriptor.ImplementationType != null ? descriptor.ImplementationType.FullName.Contains("ApplicationInsights", StringComparison.OrdinalIgnoreCase) : false));
+                Assert.Empty(services.Where(descriptor =>
+                    descriptor.ImplementationType != null ? descriptor.ImplementationType.FullName.Contains("OpenTelemetry", StringComparison.OrdinalIgnoreCase) : false));
```
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web.UnitTests/StartupTests.cs]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web/AzureMonitorOpenTelemetryLogEnricher.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web/AzureMonitorOpenTelemetryLogEnricher.cs]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web.UnitTests/AzureMonitorOpenTelemetryLogEnricherTests.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web.UnitTests/AzureMonitorOpenTelemetryLogEnricherTests.cs:51]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web/AzureMonitorOpenTelemetryLogEnricher.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web/AzureMonitorOpenTelemetryLogEnricher.cs]: Fixed.

File: src/Microsoft.Health.Fhir.Shared.Web/appsettings.json [hunk 16]
```diff
@@ -216,7 +216,9 @@
         "httpMetrics": true,
         "systemMetrics": true
     },
-    "AzureMonitor": {
+    "Telemetry": {
+        "Provider": "ApplicationInsights",
+        "InstrumentationKey": "",
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Web/appsettings.json:221]: We'll need to update the template too:

https://github.com/microsoft/fhir-server/blob/cc5576f4a90ead59f5d5b15341d5a76b754da506/samples/templates/default-azuredeploy-docker.json#L361

Will existing deployments also break?

Comment (v-iyamauchi) [reply src/Microsoft.Health.Fhir.Shared.Web/appsettings.json:221]: The PR pipeline with the template change passed. I am resolving the conversation. Please let me know if more testing needs to be done. Thank you.

**Review by brendankowitz — APPROVED** [review]

**Pull request merged**